sympy/ntheory/ecm.py

```diff
@@ -0,0 +1,309 @@
+from sympy.ntheory import sieve, isprime
+from sympy.core.power import integer_log
+from sympy.core.compatibility import as_int
+import random
+
+#----------------------------------------------------------------------------#
+#                                                                            #
+#                   Lenstra's Elliptic Curve Factorization                   #
+#                                                                            #
+#----------------------------------------------------------------------------#
+
+
+class Point:
+    """Montgomery form of Points in an elliptic curve.
+    In this form, the addition and doubling of points
+    does not need any y-coordinate information thus
+    decreasing the number of operations.
+    Using Montgomery form we try to perform point addition
+    and doubling in least amount of multiplications.
+
+    References
+    ==========
+
+    .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+    """
+
+    def __init__(self, x_cord, z_cord, a_24, mod):
+        """
+        Initial parameters for the Point class.
+
+        Parameters:
+        ===========
+
+        x_cord : X coordinate of the Point
+        z_cord : Z coordinate of the Point
+        a_24 : Parameter of the elliptic curve in Montgomery form
+        mod : modulus
+        """
+        self.x_cord = x_cord
+        self.z_cord = z_cord
+        self.a_24 = a_24
+        self.mod = mod
+
+    def __eq__(self, other):
+        from sympy import mod_inverse
+        if self.a_24 != other.a_24 or self.mod != other.mod:
+            return False
+        return self.x_cord * mod_inverse(self.z_cord, self.mod) % self.mod == other.x_cord * mod_inverse(other.z_cord, self.mod) % self.mod
+
+    def add(self, Q, diff):
+        """
+        Add two points P and Q where diff = P - Q. Moreover the assumption
```

Comment:
I don't get this, is `diff` really the difference between two points, or some random point (as shown in the example) if so how it is defined and why can't it be calculated within the method itself? and if not then the name is ambiguous.

Also using `P` is not obvious, we can replace it with `self`.

Reply:
yes,   diff is the difference between the 2 points

Reply:
And in the example i am not using diff as a random point. I have taken a initial point p1, then p2 is actually 2p1. Then p3 = p2 + p1, here the diff = p2-p1 = 2p1 - p1 = p1,
so p2.add(p1, diff=p1) gives p3.

Reply:
In that case, we can calculate it in the method itself, isn't it?

Reply:
No, we cannot. the  useful of `add` is that we can add 2 points   using less multiplications if we already know the difference between them.

Reply:
in the example i made the point such that it is `2*p1`. But  in general we don't know about this. You can observe that whereever i use `add` we knew what the difference of the points was. For example, in the `mont_ladder` it is constructed such a way that the difference  between any twp intermediate points was the initial point.

Reply:
Alright! then perhaps we need to mention this in the docstring as well that why `diff` is not computed inside the method.

Reply:
I will do that.

Reply:
I am having doubt, the diff is already known to us, so it seems conflicting      to mention that diff cannot be calculated inside the function.

Reply:
>  it seems conflicting to mention that diff cannot be calculated inside the function.

I meant to reason out "why" we are not calculating `diff` within the method (which is not obvious in general) --> so as to limit the number of multiplication operations, as you mentioned above.

Reply:
> the useful of add is that we can add 2 points using less multiplications if we already know the difference between them.

A more convincing reason for this is mentioned int the`Gaj.pdf`, below Algorithm 3, states

> The algorithm is constructed in such a way that the difference between the intermediate points P and Q, P −Q,
is always constant, and equal to the value of the initial point P0.

This makes things more clear on taking `diff` as a parameter because we would always know the `diff`. This I guess needs to be mentioned in the docstring.